File: packages/python/examples/06_task_handle_discovery.py
"""
Example: TaskHandle and Agent Card Discovery

Demonstrates:
1. Using discover=True to fetch a remote agent's card before delegating
2. Using return_handle=True to get a TaskHandle with task_id metadata
3. Using get_remote_task() to check the status of a completed task

Run:
    # Terminal 1: Start the DataAgent
    # Terminal 2: Start the OrchestratorAgent
    python examples/06_task_handle_discovery.py data
    python examples/06_task_handle_discovery.py orchestrator
"""

import sys

from a2a_lite import Agent, AgentNetwork, get_remote_task


def create_data_agent():
    """A data agent that provides lookup results."""
    agent = Agent(
        name="DataAgent",
        description="Provides data lookups for various topics",
    )

    @agent.skill("lookup", description="Look up data for a given topic")
    async def lookup(topic: str) -> dict:
        # Simulated data store
        data = {
            "weather": {"temp": "72F", "condition": "Sunny"},
            "stocks": {"AAPL": 185.50, "GOOG": 141.20},
            "news": {"headline": "AI agents are everywhere"},
        }
        return data.get(topic, {"error": f"No data for '{topic}'"})

    return agent


def create_orchestrator_agent():
    """An orchestrator that discovers the DataAgent before delegating."""
    network = AgentNetwork()
    network.add("data", "http://localhost:8787")

    agent = Agent(
        name="OrchestratorAgent",
        description="Orchestrates calls to the DataAgent with discovery and task handles",
        network=network,
    )

    @agent.skill("research", description="Research a topic using the DataAgent")
    async def research(topic: str) -> dict:
        # Step 1: Delegate with discover=True to validate the skill exists on the
        # remote agent before calling it, and return_handle=True to get a TaskHandle
        # with the task_id for later reference.
        handle = await agent.delegate(
            "data",
            "lookup",
            discover=True,
            return_handle=True,
            topic=topic,
        )

        # Step 2: The TaskHandle wraps the result and includes the task_id
        print(f"Task ID: {handle.task_id}")
        print(f"Agent URL: {handle.agent_url}")
        print(f"Result:  {handle.result}")

        # Step 3: Use the handle's convenience methods to check task status
        # (This queries the A2A protocol's tasks/get endpoint.)
        try:
            status = await handle.get_status()
            print(f"Task status (via handle): {status}")
        except Exception as e:
            # The remote agent may not support tasks/get; that's fine
            print(f"Could not fetch task status via handle: {e}")

        # Step 4: Or use the network's convenience methods
        try:
            status = await network.get_task("data", handle.task_id)
            print(f"Task status (via network): {status}")
        except Exception as e:
            print(f"Could not fetch task status via network: {e}")

        # Step 5: Cancel if needed (commented out to avoid canceling a completed task)
        # await handle.cancel()
        # await network.cancel_task("data", handle.task_id)

        return {
            "topic": topic,
            "task_id": handle.task_id,
            "data": handle.result,
        }

    return agent


if __name__ == "__main__":
    role = sys.argv[1] if len(sys.argv) > 1 else "orchestrator"

    if role == "data":
        create_data_agent().run(port=8787)
    else:
        create_orchestrator_agent().run(port=8788)

File: packages/typescript/examples/06_task_handle_discovery.ts
/**
 * Task Handle & Agent Card Discovery
 *
 * Demonstrates two features:
 *   1. Agent card discovery — inspect a remote agent's skills before calling it
 *   2. TaskHandle — get a handle to a remote task with its ID for later reference
 *
 * Architecture:
 *   - DataAgent (port 8787): exposes "lookup" and "stats" skills
 *   - OrchestratorAgent (port 8788): discovers DataAgent, delegates with handles
 *
 * Usage:
 *   # Terminal 1 — start the data agent
 *   npx tsx examples/06_task_handle_discovery.ts data
 *
 *   # Terminal 2 — start the orchestrator
 *   npx tsx examples/06_task_handle_discovery.ts
 *
 * Then call the orchestrator:
 *   curl -X POST http://localhost:8788 \
 *     -H "Content-Type: application/json" \
 *     -d '{"jsonrpc":"2.0","method":"message/send","id":"1","params":{"message":{"role":"user","parts":[{"kind":"text","text":"{\"skill\":\"query\",\"params\":{\"key\":\"users\"}}"}],"messageId":"m1"}}}'
 */
import {
  Agent,
  AgentNetwork,
  TaskHandle,
  discoverAgent,
} from '../src';

// ---------------------------------------------------------------------------
// DataAgent (port 8787) — a simple data-serving agent
// ---------------------------------------------------------------------------

function createDataAgent(): Agent {
  const agent = new Agent({
    name: 'DataAgent',
    description: 'Serves data lookups and statistics',
  });

  agent.skill('lookup', async ({ key }: { key: string }) => {
    const data: Record<string, unknown> = {
      users: [{ id: 1, name: 'Alice' }, { id: 2, name: 'Bob' }],
      products: [{ id: 1, name: 'Widget', price: 9.99 }],
    };
    return data[key] ?? [];
  });

  agent.skill('stats', async () => {
    return { uptime: '4h 32m', requests: 1042, errors: 3 };
  });

  return agent;
}

// ---------------------------------------------------------------------------
// OrchestratorAgent (port 8788) — discovers DataAgent, delegates with handles
// ---------------------------------------------------------------------------

function createOrchestratorAgent(): Agent {
  const network = new AgentNetwork();
  network.add('data', 'http://localhost:8787');

  const orchestrator = new Agent({
    name: 'OrchestratorAgent',
    description: 'Queries data agent with discovery and task handles',
    network,
  });

  // "query" skill: discovers the data agent, validates the skill, delegates
  orchestrator.skill('query', async ({ key }: { key: string }) => {
    // Step 1: Discover the DataAgent's capabilities
    const card = await discoverAgent('http://localhost:8787');
    console.log(`Discovered agent: ${card.name} v${card.version}`);
    console.log(`  Skills: ${card.skills.map((s) => s.id).join(', ')}`);
    console.log(`  Streaming: ${card.supportsStreaming}`);

    // Step 2: Delegate with returnHandle to get a TaskHandle back
    const handle = await orchestrator.delegate('data', 'lookup', { key }, {
      discover: true,
      returnHandle: true,
    }) as TaskHandle;

    console.log(`Task ID: ${handle.taskId}`);
    console.log(`Result: ${handle.toString()}`);

    // Step 3: Check task status via the handle's convenience method
    try {
      const status = await handle.getStatus();
      console.log('Status (via handle):', JSON.stringify(status, null, 2));
    } catch (err) {
      console.log('tasks/get not supported by this agent (expected for simple agents)');
    }

    // Step 4: Or check via the network by agent name
    try {
      const status2 = await network.getTask('data', handle.taskId);
      console.log('Status (via network):', JSON.stringify(status2, null, 2));
    } catch (err) {
      console.log('tasks/get via network not supported (expected for simple agents)');
    }

    // Cancel if needed:
    // await handle.cancel();
    // await network.cancelTask('data', handle.taskId);

    return {
      taskId: handle.taskId,
      data: handle.result,
      agent: card.name,
    };
  });

  return orchestrator;
}

// ---------------------------------------------------------------------------
// Main: select agent based on CLI argument
// ---------------------------------------------------------------------------

const role = process.argv[2] ?? 'orchestrator';

switch (role) {
  case 'data': {
    const agent = createDataAgent();
    agent.run({ port: 8787 });
    break;
  }
  default: {
    const agent = createOrchestratorAgent();
    agent.run({ port: 8788 });
    break;
  }
}
